### PR TITLE
Stronghold Storage Implementation

### PR DESCRIPTION
**Pull request by abdulmth**
# Description of change
Stronghold storage implementation.

Still to do:
* [x] Implement `KeyIdStorage`
* [x] Feature gate stronghold.

## Links to any relevant issues
Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

### CODE REVIEW

File: identity_storage/src/key_storage/stronghold.rs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    let public_key_procedure = iota_stronghold::procedures::PublicKey {
+      ty: keytype,
+      private_key: location,
+    };
+
+    let procedure_result = client
+      .execute_procedure(StrongholdProcedure::PublicKey(public_key_procedure))
+      .map_err(|err| {
+        KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+          .with_custom_message("stronghold procedure failed")
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:80]:
```suggestion
          .with_custom_message("stronghold public key procedure failed")
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+      private_key: location,
+      msg: data.to_vec(),
+    };
+
+    let stronghold = self.get_stronghold().await;
+    let client = get_client(&stronghold)?;
+
+    let signature: [u8; 64] = client.execute_procedure(procedure).map_err(|err| {
+      KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+        .with_custom_message("stronghold procedure failed")
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:188]:
```suggestion
        .with_custom_message("stronghold Ed25519Sign procedure failed")
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    Ok(client) => Ok(client),
+    Err(ClientError::ClientDataNotPresent) => stronghold
+      .create_client(CLIENT_PATH)
+      .map_err(|err| KeyStorageError::new(KeyStorageErrorKind::Unspecified).with_source(err)),
+    Err(err) => Err(KeyStorageError::new(KeyStorageErrorKind::Unspecified).with_source(err)),
+  }
+}
+
+async fn presist_changes(
+  secreat_manager: &StrongholdSecretManager,
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:266]:
```suggestion
  secret_manager: &StrongholdSecretManager,
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  stronghold: MutexGuard<'_, Stronghold>,
+) -> KeyStorageResult<()> {
+  stronghold.write_client(CLIENT_PATH).map_err(|err| {
+    KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+      .with_custom_message("stronghold client error")
+      .with_source(err)
+  })?;
+  // Must be dropped since `write_stronghold_snapshot` requiers the stronghold instance.
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:274]:
```suggestion
  // Must be dropped since `write_stronghold_snapshot` needs to acquire the stronghold lock.
```

File: identity_storage/src/utils/fs.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+pub fn ensure_directory<P>(path: &P) -> Result<(), std::io::Error>
+where
+  P: AsRef<Path> + ?Sized,
+{
+  if let Some(parent) = path.as_ref().parent() {
+    fs::create_dir_all(parent)?;
+  }
+
+  Ok(())
+}
```
Comment (PhilippGackstatter) [inline identity_storage/src/utils/fs.rs:16]: Is that used anywhere?

**Review by PhilippGackstatter — APPROVED** [review]
Looks good to me. 
We have to expose a constant key type for users to use and please annotate types in the storage implementations.

File: identity_storage/src/key_id_storage/tests/stronghold.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    .delete_key_id(&method_digest)
+    .await
+    .expect("deletion failed");
+
+  drop(secret_manager);
+
+  let secret_manager = StrongholdSecretManager::builder().password(PASS).build(&file).unwrap();
+  let retrieval_result: KeyIdStorageResult<KeyId> = secret_manager.get_key_id(&method_digest).await;
+  let _expected_error: KeyIdStorageError = KeyIdStorageError::new(KeyIdStorageErrorKind::KeyIdNotFound);
+  assert!(matches!(retrieval_result.unwrap_err(), _expected_error));
```
[the remark this answers is not in the document]
Comment (PhilippGackstatter) [reply identity_storage/src/key_id_storage/tests/stronghold.rs:63]: The `matches!` macro must be used differently, actually. It should contain the pattern to match against directly, not a binding.

```suggestion
  assert!(matches!(retrieval_result.unwrap_err(), KeyIdStorageError::new(KeyIdStorageErrorKind::KeyIdNotFound)));
```

[the remark this answers is not in the document]
Comment (abdulmth) [reply identity_storage/src/key_id_storage/tests/stronghold.rs]: You're right, I fixed it by matching the `KeyIdStorageErrorKind`.

File: identity_storage/Cargo.toml [hunk 7]
```diff
@@ -21,12 +21,15 @@ identity_document = { version = "=0.7.0-alpha.6", path = "../identity_document",
 identity_iota_core = { version = "=0.7.0-alpha.6", path = "../identity_iota_core", default-features = false, optional = true }
 identity_verification = { version = "=0.7.0-alpha.6", path = "../identity_verification", default_features = false }
 iota-crypto = { version = "0.23", default-features = false, features = ["blake2b", "ed25519", "random"], optional = true }
+iota-sdk = { version = "1.0", default-features = false, features = ["tls", "client", "stronghold"], optional = true }
+iota_stronghold = { version = "2.0", optional = true }
```
Comment (PhilippGackstatter) [inline identity_storage/Cargo.toml:25]: Let's please follow our codebase convention and use `default-features = false` and activate what we need.

File: identity_storage/src/key_storage/stronghold.rs [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+use std::ops::Deref;
+use std::str::FromStr;
+use tokio::sync::MutexGuard;
+
+const ED25519_KEY_TYPE_STR: &str = "Ed25519";
+static IDENTITY_VAULT_PATH: &str = "iota_identity_vault";
+pub(crate) static IDENTITY_CLIENT_PATH: &[u8] = b"iota_identity_client";
+
+/// The Ed25519 key type.
+#[allow(dead_code)]
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:39]: If this is correctly exported, the lint should not be necessary, right? Why is this here again?

File: identity_storage/src/key_storage/stronghold.rs [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+impl JwkStorage for SecretManagerWrapper {
+  async fn generate(&self, key_type: KeyType, alg: JwsAlgorithm) -> KeyStorageResult<JwkGenOutput> {
+    let stronghold = self.get_stronghold().await;
+
+    let client = get_client(&stronghold)?;
+    let key_type = StrongholdKeyType::try_from(&key_type)?;
+    check_key_alg_compatibility(key_type, alg)?;
+
+    let keytype: ProceduresKeyType = match key_type.to_string().to_lowercase().as_str() {
+      "ed25519" => Ok(ProceduresKeyType::Ed25519),
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:53]: We have the `StrongholdKeyType` enum to make this case easy to handle. Not sure why we need to lowercase convert anything here and cause another allocation.

```suggestion
    let keytype: ProceduresKeyType = match key_type {
      StrongholdKeyType::Ed25519 => Ok(ProceduresKeyType::Ed25519),
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+      key_id.to_string().as_bytes().to_vec(),
+    );
+    let stronghold = self.get_stronghold().await;
+    let client = get_client(&stronghold)?;
+    client
+      .vault(IDENTITY_VAULT_PATH.as_bytes())
+      .write_secret(location, zeroize::Zeroizing::from(secret_key.to_bytes().to_vec()))
+      .map_err(|err| {
+        KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+          .with_custom_message("stronghold client error")
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:137]:
```suggestion
          .with_custom_message("stronghold write secret failed")
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+  KeyId::new(rand::distributions::Alphanumeric.sample_string(&mut rand::thread_rng(), 32))
+}
+
+/// Check that the key type can be used with the algorithm.
+fn check_key_alg_compatibility(key_type: StrongholdKeyType, alg: JwsAlgorithm) -> KeyStorageResult<()> {
+  match (key_type, alg) {
+    (StrongholdKeyType::Ed25519, JwsAlgorithm::EdDSA) => Ok(()),
+    (key_type, alg) => Err(
+      KeyStorageError::new(crate::key_storage::KeyStorageErrorKind::KeyAlgorithmMismatch)
+        .with_custom_message(format!("`cannot use key type `{key_type}` with algorithm `{alg}`")),
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:246]:
```suggestion
        .with_custom_message(format!("cannot use key type `{key_type}` with algorithm `{alg}`")),
```

File: identity_storage/src/key_storage/stronghold.rs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+        })?;
+    }
+    _ => {
+      return Err(
+        KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+          .with_custom_message("secret manager is not of type stronghold"),
+      )
+    }
+  };
+  // .unwrap()
```
Comment (PhilippGackstatter) [inline identity_storage/src/key_storage/stronghold.rs:300]:
```suggestion
```

File: identity_storage/src/secret_manager_wrapper.rs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+use iota_sdk::client::secret::stronghold::StrongholdSecretManager;
+use iota_sdk::client::secret::SecretManager;
+use iota_stronghold::Stronghold;
+use std::sync::Arc;
+use tokio::sync::MutexGuard;
+
+/// Wrapper around `SecretManager` that implements the storage interfaces.
+#[derive(Clone)]
+pub struct SecretManagerWrapper(Arc<SecretManager>);
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:12]: This is a type that (conceptually) wraps `StrongholdSecretManager`, not `SecretManager` in general, so this should be reflected in the type's name. I would suggest `StrongholdStorage` or `StrongholdSecretStorage`, or something along those lines. Don't forget to rename the file too.

File: identity_storage/src/secret_manager_wrapper.rs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+impl SecretManagerWrapper {
+  /// Creates a new [`SecretManagerWrapper`].
+  pub fn new(stronghold_secret_manager: StrongholdSecretManager) -> Self {
+    Self(Arc::new(SecretManager::Stronghold(stronghold_secret_manager)))
+  }
+
+  /// Shared reference to the inner [`SecretManager`].
+  pub fn inner(&self) -> Arc<SecretManager> {
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:21]:
```suggestion
  pub fn as_secret_manager(&self) -> Arc<SecretManager> {
```

Follow rust convention with `as_` and be more specific about what this fn does.

File: identity_storage/src/secret_manager_wrapper.rs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    self.0.clone()
+  }
+
+  ///Acquire lock of the inner [`Stronghold`].
+  pub async fn get_stronghold(&self) -> MutexGuard<'_, Stronghold> {
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:26]:
```suggestion
  pub(crate) async fn get_stronghold(&self) -> MutexGuard<'_, Stronghold> {
```

This can be `pub(crate)` if I'm not mistaken.

File: identity_storage/src/secret_manager_wrapper.rs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  pub fn new(stronghold_secret_manager: StrongholdSecretManager) -> Self {
+    Self(Arc::new(SecretManager::Stronghold(stronghold_secret_manager)))
+  }
+
+  /// Shared reference to the inner [`SecretManager`].
+  pub fn inner(&self) -> Arc<SecretManager> {
+    self.0.clone()
+  }
+
+  ///Acquire lock of the inner [`Stronghold`].
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:25]:
```suggestion
  /// Acquire lock of the inner [`Stronghold`].
```

File: identity_storage/src/secret_manager_wrapper.rs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use iota_sdk::client::secret::stronghold::StrongholdSecretManager;
+use iota_sdk::client::secret::SecretManager;
+use iota_stronghold::Stronghold;
+use std::sync::Arc;
+use tokio::sync::MutexGuard;
+
+/// Wrapper around `SecretManager` that implements the storage interfaces.
+#[derive(Clone)]
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:11]: Can we derive `Debug` as well?

File: identity_storage/src/secret_manager_wrapper.rs [hunk 18]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2020-2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use iota_sdk::client::secret::stronghold::StrongholdSecretManager;
+use iota_sdk::client::secret::SecretManager;
+use iota_stronghold::Stronghold;
+use std::sync::Arc;
+use tokio::sync::MutexGuard;
+
+/// Wrapper around `SecretManager` that implements the storage interfaces.
```
Comment (PhilippGackstatter) [inline identity_storage/src/secret_manager_wrapper.rs:10]:
```suggestion
/// Wrapper around a `StrongholdSecretManager` that implements the [`KeyIdStorage`](todo:link) and [`JwkStorage`](todo:link) interfaces.
```

Yes, technically it wraps a `SecretManager`, but conceptually it's a `StrongholdSecretManager` and that is what matters to the user.

**Review by PhilippGackstatter — CHANGES_REQUESTED** [review]

**Comment by abdulmth** [comment]
Maybe we can change the examples in a separate PR.

File: identity_storage/Cargo.toml [hunk 19]
```diff
@@ -40,4 +43,6 @@ memstore = ["dep:tokio", "dep:rand", "dep:iota-crypto", "identity_verification/e
 # Enables `Send` + `Sync` bounds for the storage traits.
 send-sync-storage = []
 # Implements the JwkStorageDocumentExt trait for IotaDocument
+# Exposes the stronghold implementations for the storage traits.
+stronghold = ["dep:iota-sdk", "dep:iota_stronghold", "dep:zeroize", "dep:rand", "dep:iota-crypto"]
```
Comment (PhilippGackstatter) [inline identity_storage/Cargo.toml:47]: This causes errors.

```
cargo check --no-default-features -p identity_storage --features stronghold
```

File: identity_storage/src/lib.rs [hunk 20]
```diff
@@ -17,7 +17,14 @@
 pub mod key_id_storage;
 pub mod key_storage;
 pub mod storage;
+/// Wrapper around `SecretManager` that implements the storage interfaces.
```
Comment (PhilippGackstatter) [inline identity_storage/src/lib.rs:20]: Modules are conventionally documented inside the `mod.rs` file with a `//!` comment.

File: identity_storage/src/stronghold_storage.rs [hunk 21]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2020-2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use iota_sdk::client::secret::stronghold::StrongholdSecretManager;
+use iota_sdk::client::secret::SecretManager;
+use iota_stronghold::Stronghold;
+use std::sync::Arc;
+use tokio::sync::MutexGuard;
+
+/// Wrapper around `StrongholdSecretManager` that implements the [`KeyIdStorage`](crate::KeyIdStorage)
```
Comment (PhilippGackstatter) [inline identity_storage/src/stronghold_storage.rs:10]:
```suggestion
/// Wrapper around a [`StrongholdSecretManager`] that implements the [`KeyIdStorage`](crate::KeyIdStorage)
```

File: identity_storage/src/stronghold_storage.rs [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+pub struct StrongholdStorage(Arc<SecretManager>);
+
+impl StrongholdStorage {
+  /// Creates a new [`StrongholdStorage`].
+  pub fn new(stronghold_secret_manager: StrongholdSecretManager) -> Self {
+    Self(Arc::new(SecretManager::Stronghold(stronghold_secret_manager)))
+  }
+
+  /// Shared reference to the inner [`SecretManager`].
+  pub fn as_secret_manager(&self) -> Arc<SecretManager> {
```
Comment (PhilippGackstatter) [inline identity_storage/src/stronghold_storage.rs:22]:
```suggestion
  pub fn as_secret_manager(&self) -> &SecretManager {
```

**Review by PhilippGackstatter — APPROVED** [review]

File: identity_storage/src/stronghold_storage.rs (hunk 22 above)
Comment (abdulmth) [reply identity_storage/src/stronghold_storage.rs:22]: This doesn't work, `cannot return reference to temporary value ...`

File: identity_storage/Cargo.toml (hunk 19 above)
Comment (abdulmth) [reply identity_storage/Cargo.toml]: fixed, tokio was missing

File: identity_storage/src/stronghold_storage.rs (hunk 22 above)
Comment (PhilippGackstatter) [reply identity_storage/src/stronghold_storage.rs:24]:
```
  pub fn as_secret_manager(&self) -> &SecretManager {
    self.0.as_ref()
  }
```

Replacing the function with this compiles fine for me.

Comment (abdulmth) [reply identity_storage/src/stronghold_storage.rs]: thanks, my bad, I replaced it
